qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -441,188 +453,208 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
-                    )
-
-                results = self._quantum_instance.execute(circuits, had_transpiled=True)
+        if self._enforce_psd and is_symmetric:
+            # Find the closest positive semi-definite approximation to symmetric kernel matrix.
+            # The (symmetric) matrix should always be positive semi-definite by construction,
+            # but this can be violated in case of noise, such as sampling noise, thus the
+            # adjustment is only done if NOT using the statevector simulation.
+            D, U = np.linalg.eig(kernel)  # pylint: disable=invalid-name
+            kernel = U @ np.diag(np.maximum(0, D)) @ U.transpose()
 
-                matrix_elements = [
-                    self._compute_overlap(circuit, results, is_statevector_sim, measurement_basis)
-                    for circuit in range(len(circuits))
-                ]
+        return kernel
 
-                for (i, j), value in zip(to_be_computed_index, matrix_elements):
-                    kernel[i, j] = value
-                    if is_symmetric:
-                        kernel[j, i] = kernel[i, j]
+    def _get_indices(
+        self, x_vec: np.ndarray, y_vec: np.ndarray, is_symmetric: bool
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        # get indices to calculate
+        if is_symmetric:
+            row_indices, col_indices = np.triu_indices(x_vec.shape[0], k=1)  # remove diagonal
+        else:
+            row_indices, col_indices = np.indices((x_vec.shape[0], y_vec.shape[0]))
+            row_indices = np.asarray(row_indices.flat)
+            col_indices = np.asarray(col_indices.flat)
+        return row_indices, col_indices
 
-            if self._enforce_psd and is_symmetric:
-                # Find the closest positive semi-definite approximation to symmetric kernel matrix.
-                # The (symmetric) matrix should always be positive semi-definite by construction,
-                # but this can be violated in case of noise, such as sampling noise, thus the
-                # adjustment is only done if NOT using the statevector simulation.
-                D, U = np.linalg.eig(kernel)  # pylint: disable=invalid-name
-                kernel = U @ np.diag(np.maximum(0, D)) @ U.transpose()
+    def _check_training_parameters_bound(self) -> None:
+        # Ensure all training parameters have been bound in the feature map circuit.
+        unbound_params = self.get_unbound_training_parameters()
+        if unbound_params:
+            raise ValueError(
+                f"""
+                The feature map circuit contains unbound training parameters ({unbound_params}).
+                All training parameters must be bound to numerical values before evaluating
+                the kernel matrix.
+                """
+            )
 
-        return kernel
+    def _raise_incompatible_feature_map(self, vec_name: str, vec_len: int) -> None:
```

Comment:
The return type here ought to be `NoReturn` I think since all this function never ever returns